agent/xds/routes.go

```diff
@@ -11,41 +12,128 @@ import (
+// For any HTTP service we will return a default route.
+func (s *Server) routesFromSnapshotTerminatingGateway(_ connectionInfo, cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
 	if cfgSnap == nil {
 		return nil, errors.New("nil config given")
 	}
+	logger := s.Logger.Named(logging.TerminatingGateway)
+
+	var resources []proto.Message
+	for _, svc := range cfgSnap.TerminatingGateway.ValidServices() {
+		clusterName := connect.ServiceSNI(svc.Name, "", svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+		resolver, hasResolver := cfgSnap.TerminatingGateway.ServiceResolvers[svc]
+
+		svcConfig := cfgSnap.TerminatingGateway.ServiceConfigs[svc]
+
+		cfg, err := ParseProxyConfig(svcConfig.ProxyConfig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse upstream config: %v", err)
+		}
+		if !structs.IsProtocolHTTPLike(cfg.Protocol) {
+			// Routes can only be defined for HTTP services
+			continue
+		}
+
+		if !hasResolver {
+			// Use a zero value resolver with no timeout and no subsets
+			resolver = &structs.ServiceResolverConfigEntry{}
+		}
+
+		var lb *structs.EnvoyLBConfig
+		if resolver.LoadBalancer != nil {
+			lb = resolver.LoadBalancer.EnvoyConfig
+		}
+		route, err := makeNamedDefaultRouteWithLB(clusterName, lb)
+		if err != nil {
+			logger.Error("failed to make route", "cluster", clusterName, "error", err)
+			continue
+		}
+		resources = append(resources, route)
+
+		// If there is a service-resolver for this service then also setup routes for each subset
+		for name := range resolver.Subsets {
+			clusterName = connect.ServiceSNI(svc.Name, name, svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+			route, err := makeNamedDefaultRouteWithLB(clusterName, lb)
+			if err != nil {
+				logger.Error("failed to make route", "cluster", clusterName, "error", err)
+				continue
+			}
+			resources = append(resources, route)
+		}
+	}
+
+	return resources, nil
+}
+
+func makeNamedDefaultRouteWithLB(clusterName string, lb *structs.EnvoyLBConfig) (*envoy.RouteConfiguration, error) {
+	action := makeRouteActionFromName(clusterName)
+
+	if err := injectLBToRouteAction(lb, action.Route); err != nil {
+		return nil, fmt.Errorf("failed to apply load balancer configuration to route action: %v", err)
+	}
+
+	return &envoy.RouteConfiguration{
+		Name: clusterName,
+		VirtualHosts: []*envoyroute.VirtualHost{
+			{
+				Name:    clusterName,
+				Domains: []string{"*"},
+				Routes: []*envoyroute.Route{
+					{
+						Match:  makeDefaultRouteMatch(),
+						Action: action,
+					},
+				},
+			},
+		},
+		// ValidateClusters defaults to true when defined statically and false
+		// when done via RDS. Re-set the sane value of true to prevent
+		// null-routing traffic.
+		ValidateClusters: makeBoolValue(true),
+	}, nil
+}
+
+// routesFromSnapshotConnectProxy returns the xDS API representation of the
+// "routes" in the snapshot.
+func routesForConnectProxy(
```

Comment:
(nit)
I have a suspicion that if you flipped the order of this method and the prior one in the delta that the overall code diff would make the PR read a little easier, as the connect-proxy method before and after would diff out together, instead of the weird splice-diff that is happening due to the line-overlap with the connect-proxy one and the TG one.

Reply:
Will try it out